mttl/models/modifiers/sparse_mask.py

```diff
@@ -116,67 +172,180 @@ def get_regular_sparse_mask(m):
+    elif parameter_selection_procedure=='model':
+        # b.1 compute score
         num_params_to_keep = 0
         grads = []
         for m in module.modules():
             if isinstance(m, SparseMaskModule):
-                assert (
-                    m.sparse_cat == "regular_sparse"
-                ), "parameter_selection_procedure over `model` is not implemented for `block_sparse`"
-                num_params_to_keep += int(
-                    torch.numel(m.sparse_layer.weight_mask) * m.keep_ratio
-                )
+                assert m.sparse_cat == 'regular_sparse',  "parameter_selection_procedure over `model` is not implemented for `block_sparse`"
+                num_params_to_keep += int(torch.numel(m.sparse_layer.weight_mask)*m.keep_ratio)
                 grads.append(m.sparse_layer.weight_mask.grad.flatten().cpu())
-
-        threshold, _ = torch.topk(
-            torch.stack(grads).flatten(), num_params_to_keep, sorted=True
-        )
+        threshold, _ = torch.topk(torch.stack(grads).flatten(), num_params_to_keep, sorted=True)
         accepted_score = threshold[-1]
         # b.2 mask
         for m in module.modules():
             if isinstance(m, SparseMaskModule):
-                keep_masks = (m.sparse_layer.weight_mask.grad >= accepted_score).float()
+                keep_masks = (m.sparse_layer.weight_mask.grad>=accepted_score).float()
+                if print_statement: print('sparsity', (keep_masks.sum()/m.sparse_layer.weight_mask.numel())*100, 'expected', m.keep_ratio*100)
                 m.revert_weight_grad_and_update_mask(keep_masks)
 
+    # drop layer and params
+    elif parameter_selection_procedure=='layer_and_param':
+        num_params_to_keep = 0
+        grads = []
+        for m in module.modules():
+            if isinstance(m, SparseMaskModule):
+                assert m.sparse_cat == 'regular_sparse',  "parameter_selection_procedure over `model` is not implemented for `block_sparse`"
+                num_params_to_keep += int(torch.numel(m.sparse_layer.weight_mask)*m.keep_ratio)
+                grads.append(m.sparse_layer.weight_mask.grad.flatten().cpu())
+        threshold, _ = torch.topk(torch.stack(grads).flatten(), num_params_to_keep, sorted=True)
+        accepted_score = threshold[-1]
+        # b.2 mask
+        for m in module.modules():
+            if isinstance(m, SparseMaskModule):
+                keep_masks = (m.sparse_layer.weight_mask.grad>=accepted_score).float()
+                if  (keep_masks.sum()/m.sparse_layer.weight_mask.numel())>= m.keep_ratio:
+                    keep_masks = get_regular_sparse_mask(m)
+                    m.revert_weight_grad_and_update_mask(keep_masks)
+                else:
+                    keep_masks = torch.zeros_like(m.sparse_layer.weight)
+                    m.revert_weight_grad_and_update_mask(keep_masks)
+    # based on gradient-magnitude
+    elif parameter_selection_procedure=='gradient_magnitude':
```

Comment:
can you add tests for the new permutations of these parameters in `test_sparse_mask.py`?